async-nats/src/connector.rs

```diff
@@ -0,0 +1,249 @@
+use crate::connection::Connection;
+use crate::tls;
+use crate::Authorization;
+use crate::ClientOp;
+use crate::ConnectInfo;
+use crate::Protocol;
+use crate::ServerAddr;
+use crate::ServerInfo;
+use crate::ServerOp;
+use crate::SocketAddr;
+use crate::ToServerAddrs;
+use crate::LANG;
+use crate::VERSION;
+use bytes::BytesMut;
+use std::cmp;
+use std::collections::HashMap;
+use std::io;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::io::BufWriter;
+use tokio::io::ErrorKind;
+use tokio::net::TcpStream;
+use tokio::time::sleep;
+use tokio_rustls::rustls::{self};
+
+pub(crate) struct ConnectorOptions {
+    pub(crate) tls_required: bool,
+    pub(crate) certificates: Vec<PathBuf>,
+    pub(crate) client_cert: Option<PathBuf>,
+    pub(crate) client_key: Option<PathBuf>,
+    pub(crate) tls_client_config: Option<rustls::ClientConfig>,
+    pub(crate) auth: Authorization,
+    pub(crate) no_echo: bool,
+}
+
+/// Maintains a list of servers and establishes connections.
+pub(crate) struct Connector {
+    /// A map of servers and number of connect attempts.
+    servers: HashMap<ServerAddr, usize>,
+    options: ConnectorOptions,
+}
+
+impl Connector {
+    pub(crate) fn new<A: ToServerAddrs>(
+        addrs: A,
+        options: ConnectorOptions,
+    ) -> Result<Connector, io::Error> {
+        let servers = addrs
+            .to_server_addrs()?
+            .into_iter()
+            .map(|addr| (addr, 0))
+            .collect();
+
+        Ok(Connector { servers, options })
+    }
+
+    pub(crate) async fn connect(&mut self) -> Result<(ServerInfo, Connection), io::Error> {
+        for _ in 0..128 {
+            if let Ok(inner) = self.try_connect().await {
+                return Ok(inner);
+            }
+        }
+
+        Err(io::Error::new(io::ErrorKind::Other, "unable to connect"))
+    }
+
+    pub(crate) async fn try_connect(&mut self) -> Result<(ServerInfo, Connection), io::Error> {
```

Comment:
ServerInfo is actually never used.

Let's return just Connection instead of tuple.